qiskit/providers/aer/noise/utils/idle_scheduler.py

```diff
@@ -0,0 +1,100 @@
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Idle gates scheduler module
+
+The goal of this module is to add idle gates to a circuit instead of
+having blank spaces with no operations at times where no gate can be applied
+to the qubit (due to qubit dependenices or barriers)
+"""
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.converters import circuit_to_dag
+from qiskit.converters import dag_to_circuit
+from qiskit.extensions.standard.iden import IdGate
+
+
+def schedule_idle_gates(circuit, op_times=None, default_op_time=1):
+    """
+    This function gets a circuit and returns a new one with the idle
+    gates in place; each idle gate is labeled "id_<name>" where <name> is the name of
+    the slowest gate that runs in parallel
+
+    Args:
+        circuit (QuantumCircuit): The circuit to add idle gates to
+        op_times (dictionary): dictionary of times taken by specific gate types
+        default_op_time (float or int): default time for gates not found in op_times
+
+    Returns:
+        QuantumCircuit: The new circuit with the added idle gates
+
+    """
+    scheduler = IdleScheduler(op_times, default_op_time)
+    return scheduler.schedule(circuit)
+
+
+class IdleScheduler():
+    """Adds idle gates to given circuits"""
+    def __init__(self, op_times, default_op_time):
+        if op_times is None:
+            self.op_times = {}
+        else:
+            self.op_times = op_times
+        self.default_op_time = default_op_time
+        self.circuit = None
+        self.idle_times = None
+
+    def schedule(self, circuit):
+        """Adds the idle gates to the circuit
+            Args:
+               circuit (QuantumCircuit): The circuit to modify
+            Returns:
+               QuantumCircuit: The modified circuit
+       """
+        self.circuit = circuit
+        new_dag = DAGCircuit()
+        self.idle_times = {qubit: 0 for qubit in self.circuit.qubits}
+        dag = circuit_to_dag(circuit)
+        # convert to circuit and back as hack to prevent nondeterminism in DAGCircuit
+        layers = [circuit_to_dag(dag_to_circuit(l['graph'])) for l in dag.layers()]
+        for layer in layers:
+            new_layer_graph = self.add_identities_to_layer(layer)
+            new_dag.extend_back(new_layer_graph)
+        return dag_to_circuit(new_dag)
+
+    def add_identities_to_layer(self, layer):
+        """Adds the idle gates to a specific layer in the circuit
+            Args:
+               layer (dict): The layer to modify (contains a circuit graph)
+            Returns:
+               dict: The modified layer
+        """
+        max_op_time, max_op_name = max(
+            [(self.op_times.get(node.name, self.default_op_time), node.name)
```

Comment:
The `op_times` is a dict that is keyed on gate names. Therefore it prohibits things like "u3 on qubit 0 is 200 ns and u3 on qubit 1 is 300 ns".
Is there a clean way to make this more general? How does Aer currently handle the difference in latency on gates of the same kind?

Reply:
This should not be too difficult; key the times using a tuple `(node_name, qubit_num)`. The question is - do we need this right now (if so, I'll implement).

Reply:
Take a look at gate times for some devices to see if different physical U or CX gates have different latencies. If so, then I'd say this should match that.

Reply:
I would try and make this mimic the same format used for specifying custom gate times in the device noise model, which is a list with entries `(name, qubits, time)`. If qubits is `None` it applies to that gate on all qubits, otherwise it applies only to those qubits.

Usually devices will have the same gate time on all qubits for all single-qubit gates, but different `cx` gate times on different pairs of qubits. In the device noise model this could be entered like

```python
gate_times = [('u1', None, 0), ('u2', None, 100), ('u3', None, 200),
    ('cx', [0, 2], 678), ('cx', [1, 2], 547), ('cx', [2, 3], 721)]
```

Reply:
Chris, your example raises two separate questions.

1. What to do in case of conflicts, i.e. `gate_times=[('u1', None, 0), ('u1', [1], 17)]`
2. How to handle gates like `u1`: apply only to `u1` gates, or for any gate defined via such gates? How can this be checked in Aer/Terra?

Reply:
I introduced the new interface at 89d1d8f30fe0c033d842b047134b0ebb588bb8c3.

Note that for cx gates, different **order** of qubits results in "different gate", i.e. `('cx', [0, 2], 678), ('cx', [2,0], 356)` is acceptable, and `('cx', [0, 2], 678)` will give the default (1) time to `('cx', [2,0])`.

Reply:
You could do one of two options: 

1. `("u1", None, 0)` would be a default for all qubits not otherwise specified, and `("u"`, [1], 17)` would override the default for qubit-0. 
2. Raise an error if there a situation like that.

I think the first case is more useful. I'm not sure what device noise model does in this situation, but we should make them match. And for cx gates you are right that different order of qubits should be considered distinct gates.

Reply:
Glad to say option 1 is already what the code does, so I'm resolving this.